File: surf-cloud-standalone/src/main/java/dev/slne/surf/cloud/standalone/plugin/StandalonePluginProcessor.java
package dev.slne.surf.cloud.standalone.plugin;

import org.jetbrains.annotations.NotNull;
import org.springframework.aop.framework.AopInfrastructureBean;
import org.springframework.aop.framework.AopProxyUtils;
import org.springframework.beans.BeansException;
import org.springframework.beans.factory.BeanCreationException;
import org.springframework.beans.factory.config.BeanPostProcessor;
import org.springframework.core.annotation.AnnotationUtils;
import org.springframework.stereotype.Component;

@Component
public class StandalonePluginProcessor implements BeanPostProcessor {

  private final StandalonePluginManager standalonePluginManager;

  public StandalonePluginProcessor(StandalonePluginManager standalonePluginManager) {
    this.standalonePluginManager = standalonePluginManager;
  }

  @Override
  public Object postProcessAfterInitialization(@NotNull Object bean, @NotNull String beanName) throws BeansException {
    if (bean instanceof AopInfrastructureBean) {
      return bean;
    }

    final Class<?> targetClass = AopProxyUtils.ultimateTargetClass(bean);
    if (AnnotationUtils.findAnnotation(targetClass, StandalonePluginMeta.class) == null) {
      return bean;
    }

    if (!(bean instanceof StandalonePlugin plugin)) {
      throw new BeanCreationException(beanName, "Bean annotated with @StandalonePluginMeta must extend StandalonePlugin");
    }

    plugin.start();
    standalonePluginManager.addPlugin(plugin);
    return bean;
  }
}

File: surf-cloud-standalone/src/main/kotlin/dev/slne/surf/cloud/standalone/plugin/StandalonePluginProcessor.kt
package dev.slne.surf.cloud.standalone.plugin

import dev.slne.surf.cloud.api.util.findAnnotation
import dev.slne.surf.cloud.api.util.logger
import dev.slne.surf.cloud.api.util.ultimateTargetClass
import org.springframework.aop.framework.AopInfrastructureBean
import org.springframework.beans.factory.BeanCreationException
import org.springframework.beans.factory.config.BeanPostProcessor
import org.springframework.stereotype.Component

@Component
class StandalonePluginProcessor : BeanPostProcessor {
    private val log = logger()
    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any? {
        if (bean is AopInfrastructureBean) return bean

        val targetClass = bean.ultimateTargetClass()
        targetClass.findAnnotation<StandalonePluginMeta>() ?: return bean

        if (bean !is StandalonePlugin) {
            log.atSevere()
                .withCause(
                    BeanCreationException(
                        beanName,
                        "Bean annotated with @StandalonePluginMeta must extend StandalonePlugin"
                    )
                )
                .log("Bean annotated with @StandalonePluginMeta must extend StandalonePlugin")

            return null
        }

        try {
            bean.start()
        } catch (e: Exception) {
            log.atSevere()
                .withCause(e)
                .log("Failed to start plugin '%s'", bean.id)
            return null
        }


        StandalonePluginManager.addPlugin(bean)
        return bean
    }
}
